docs/Chipyard-Basics/Debugging-RTL.rst

```diff
@@ -0,0 +1,91 @@
+Debugging RTL
+======================
+
+While the packaged Chipyard configs and RTL have been tested to work,
+users will typically want to build custom chips by adding their own
+IP, or by modifying existing Chisel generators. Such changes might introduce
+bugs. This section aims to run through a typical debugging flow
+using Chipyard. We assume the user has a custom SoC configuration,
+and is trying to verify functionality by running some software test.
+
+Bringup Method
+---------------------------
+
+There are two approaches for running a custom application binary on a
+Chipyard. TSI uses a Test Serial Interface to write the binary into
+simulated memory, while DTM uses the SoC's debug module to execute a program
+that fetches instruction bits from the host.
+
+The DTM is used in actual chip bringup, while the TSI in simulation is
+designed to fast and deterministic. For this reason we generally prefer
+TSI for initial debugging of SW simulation.
+
+Waveforms
+---------------------------
+
+The default SW simulators do not dump waveforms during execution. To build
+simulators with wave dump capabilities use must use the ``debug`` make target.
+For example:
+
+.. code-block:: shell
+
+   make CONFIG=CustomConfig debug
+
+The ``run-binary-debug`` rule will also automatically build a simulator,
+run it on a custom binary, and generate a waveform. For example, to run a
+test on ``helloworld.riscv``, use
+
+.. code-block:: shell
+
+   make CONFIG=CustomConfig run-binary-debug BINARY=helloworld.riscv
```

Comment:
IMO it's better to do ``make CONFIG=... BINARY=... OTHER=... target``. However, I don't know if we want to have this be a standard in the docs... if not then this can stay this way.

Reply:
It doesn't really matter. This way minimizes cursor movement when running many different binaries.